File: cases/examples_sergey/analytics_csv/scripts/generate_template.py
"""Generate sales_report_template.xlsx with 5 sheets and charts (dev-only)."""
from __future__ import annotations

from pathlib import Path

from openpyxl import Workbook
from openpyxl.chart import BarChart, LineChart, PieChart, Reference

OUT = Path(__file__).resolve().parent.parent / "analytics_csv" / "templates" / "sales_report_template.xlsx"


def add_bar(ws, title: str, cat_min_row: int, cat_max_row: int, val_col: int, anchor: str = "D1"):
    chart = BarChart()
    chart.type = "col"
    chart.title = title
    chart.y_axis.title = "Значение"
    cats = Reference(ws, min_col=1, min_row=cat_min_row, max_row=cat_max_row)
    data = Reference(ws, min_col=val_col, min_row=cat_min_row - 1, max_row=cat_max_row)
    chart.add_data(data, titles_from_data=True)
    chart.set_categories(cats)
    chart.height = 12
    chart.width = 18
    ws.add_chart(chart, anchor)


def add_pie(ws, title: str, min_row: int, max_row: int, anchor: str = "E1"):
    chart = PieChart()
    chart.title = title
    labels = Reference(ws, min_col=1, min_row=min_row, max_row=max_row)
    data = Reference(ws, min_col=2, min_row=min_row - 1, max_row=max_row)
    chart.add_data(data, titles_from_data=True)
    chart.set_categories(labels)
    chart.height = 12
    chart.width = 16
    ws.add_chart(chart, anchor)


def add_line(ws, title: str, min_row: int, max_row: int, anchor: str = "E1"):
    chart = LineChart()
    chart.title = title
    chart.y_axis.title = "Покупки"
    cats = Reference(ws, min_col=1, min_row=min_row, max_row=max_row)
    data = Reference(ws, min_col=2, min_row=min_row - 1, max_row=max_row)
    chart.add_data(data, titles_from_data=True)
    chart.set_categories(cats)
    chart.height = 12
    chart.width = 18
    ws.add_chart(chart, anchor)


def main() -> None:
    wb = Workbook()
    wb.remove(wb.active)

    # Сводка
    ws = wb.create_sheet("Сводка", 0)
    ws.append(["Метрика", "Значение"])
    ws.append(["Число покупок (cart)", 0])
    ws.append(["Выручка, руб.", 0])
    ws.append(["Средний чек, руб.", 0])
    ws.append(["Уникальные покупатели", 0])
    add_bar(ws, "Сводка по продажам", 2, 5, 2)

    # Бренды
    ws = wb.create_sheet("Бренды", 1)
    ws.append(["Бренд", "Покупки", "Выручка, руб.", "Доля, %"])
    for _ in range(10):
        ws.append(["", 0, 0, 0])
    add_bar(ws, "Топ брендов по покупкам", 2, 11, 2)

    # Категории
    ws = wb.create_sheet("Категории", 2)
    ws.append(["Категория", "Покупки", "Выручка, руб.", "Доля, %"])
    for _ in range(10):
        ws.append(["", 0, 0, 0])
    add_pie(ws, "Категории по покупкам", 2, 11)

    # Воронка
    ws = wb.create_sheet("Воронка", 3)
    ws.append(["Этап", "Количество", "Конверсия, %"])
    ws.append(["Просмотры (view)", 0, 100])
    ws.append(["Покупки (purchase)", 0, 0])
    add_bar(ws, "Воронка продаж", 2, 3, 2)

    # Динамика
    ws = wb.create_sheet("Динамика", 4)
    ws.append(["Дата", "Покупки", "Выручка, руб."])
    for _ in range(31):
        ws.append(["", 0, 0])
    add_line(ws, "Динамика покупок", 2, 32)

    OUT.parent.mkdir(parents=True, exist_ok=True)
    wb.save(OUT)
    print(f"Written: {OUT}")


if __name__ == "__main__":
    main()

File: cases/examples_sergey/analytics_csv/scripts/generate_template.mjs
/**
 * One-time dev script: builds sales_report_template.xlsx with 5 sheets and charts.
 * Run from cases/analytics_csv: node scripts/generate_template.mjs
 */
import ExcelJS from 'exceljs';
import { mkdir, writeFile } from 'node:fs/promises';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';

const __dirname = dirname(fileURLToPath(import.meta.url));
const outDir = join(__dirname, '..', 'analytics_csv', 'templates');
const outPath = join(outDir, 'sales_report_template.xlsx');

function addBarChart(ws, title, categories, values, anchor) {
	ws.addChart({
		type: 'bar',
		title: { name: title },
		legend: { position: 'bottom' },
		plotArea: {
			dataSeries: [
				{
					title: values.title,
					categories: { ranges: [categories] },
					values: { ranges: [values] }
				}
			]
		},
		anchor: anchor || { col: 4, row: 0, nativeCol: 4, nativeRow: 0 }
	});
}

function addLineChart(ws, title, categories, values, anchor) {
	ws.addChart({
		type: 'line',
		title: { name: title },
		legend: { position: 'bottom' },
		plotArea: {
			dataSeries: [
				{
					title: values.title,
					categories: { ranges: [categories] },
					values: { ranges: [values] }
				}
			]
		},
		anchor: anchor || { col: 4, row: 0, nativeCol: 4, nativeRow: 0 }
	});
}

function addPieChart(ws, title, categories, values, anchor) {
	ws.addChart({
		type: 'pie',
		title: { name: title },
		legend: { position: 'right' },
		plotArea: {
			dataSeries: [
				{
					title: values.title,
					categories: { ranges: [categories] },
					values: { ranges: [values] }
				}
			]
		},
		anchor: anchor || { col: 4, row: 0, nativeCol: 4, nativeRow: 0 }
	});
}

async function main() {
	const wb = new ExcelJS.Workbook();
	wb.creator = 'analytics_csv';
	wb.created = new Date();

	// Sheet 1: Сводка
	const summary = wb.addWorksheet('Сводка');
	summary.columns = [
		{ header: 'Метрика', key: 'metric', width: 28 },
		{ header: 'Значение', key: 'value', width: 18 }
	];
	summary.addRows([
		{ metric: 'Число покупок (cart)', value: 0 },
		{ metric: 'Выручка, руб.', value: 0 },
		{ metric: 'Средний чек, руб.', value: 0 },
		{ metric: 'Уникальные покупатели', value: 0 }
	]);
	addBarChart(summary, 'Сводка по продажам', "'Сводка'!$A$2:$A$5", {
		title: 'Значение',
		ranges: ["'Сводка'!$B$2:$B$5"]
	});

	// Sheet 2: Бренды
	const brands = wb.addWorksheet('Бренды');
	brands.columns = [
		{ header: 'Бренд', key: 'brand', width: 22 },
		{ header: 'Покупки', key: 'count', width: 12 },
		{ header: 'Выручка, руб.', key: 'revenue', width: 16 },
		{ header: 'Доля, %', key: 'share', width: 10 }
	];
	for (let i = 0; i < 10; i++) {
		brands.addRow({ brand: '', count: 0, revenue: 0, share: 0 });
	}
	addBarChart(brands, 'Топ брендов по покупкам', "'Бренды'!$A$2:$A$11", {
		title: 'Покупки',
		ranges: ["'Бренды'!$B$2:$B$11"]
	});

	// Sheet 3: Категории
	const categories = wb.addWorksheet('Категории');
	categories.columns = [
		{ header: 'Категория', key: 'category', width: 36 },
		{ header: 'Покупки', key: 'count', width: 12 },
		{ header: 'Выручка, руб.', key: 'revenue', width: 16 },
		{ header: 'Доля, %', key: 'share', width: 10 }
	];
	for (let i = 0; i < 10; i++) {
		categories.addRow({ category: '', count: 0, revenue: 0, share: 0 });
	}
	addPieChart(categories, 'Категории по покупкам', "'Категории'!$A$2:$A$11", {
		title: 'Покупки',
		ranges: ["'Категории'!$B$2:$B$11"]
	});

	// Sheet 4: Воронка
	const funnel = wb.addWorksheet('Воронка');
	funnel.columns = [
		{ header: 'Этап', key: 'stage', width: 22 },
		{ header: 'Количество', key: 'count', width: 14 },
		{ header: 'Конверсия, %', key: 'conversion', width: 14 }
	];
	funnel.addRows([
		{ stage: 'Просмотры (view)', count: 0, conversion: 100 },
		{ stage: 'Покупки (purchase)', count: 0, conversion: 0 }
	]);
	addBarChart(funnel, 'Воронка продаж', "'Воронка'!$A$2:$A$3", {
		title: 'Количество',
		ranges: ["'Воронка'!$B$2:$B$3"]
	});

	// Sheet 5: Динамика
	const dynamics = wb.addWorksheet('Динамика');
	dynamics.columns = [
		{ header: 'Дата', key: 'date', width: 14 },
		{ header: 'Покупки', key: 'count', width: 12 },
		{ header: 'Выручка, руб.', key: 'revenue', width: 16 }
	];
	for (let i = 0; i < 31; i++) {
		dynamics.addRow({ date: '', count: 0, revenue: 0 });
	}
	addLineChart(dynamics, 'Динамика покупок', "'Динамика'!$A$2:$A$32", {
		title: 'Покупки',
		ranges: ["'Динамика'!$B$2:$B$32"]
	});

	await mkdir(outDir, { recursive: true });
	await wb.xlsx.writeFile(outPath);
	console.log('Written:', outPath);
}

main().catch((err) => {
	console.error(err);
	process.exit(1);
});
